python/lsst/daf/butler/queries/tree/_query_tree.py

```
@@ -0,0 +1,314 @@
+from ._column_set import ColumnSet
+from ._predicate import Predicate
+
+DataCoordinateUploadKey: TypeAlias = uuid.UUID
+
+MaterializationKey: TypeAlias = uuid.UUID
+
+
+def make_identity_query_tree(universe: DimensionUniverse) -> QueryTree:
+    """Make an initial query tree with empty dimensions and a single logical
+    row.
+
+    This method should be used by `Butler._query` to construct the initial
+    query tree.  This tree is a useful initial state because it is the
+    identity for joins, in that joining any other query tree to the identity
+    yields that query tree.
+
+    Parameters
+    ----------
+    universe : `..DimensionUniverse`
+        Definitions for all dimensions.
+
+    Returns
+    -------
+    tree : `QueryTree`
+        A tree with empty dimensions.
+    """
+    return QueryTree(dimensions=universe.empty.as_group())
+
+
+@final
+class DatasetSearch(QueryTreeBase):
+    """Information about a dataset search joined into a query tree.
+
+    The dataset type name is the key of the dictionary (in `QueryTree`) where
+    this type is used as a value.
+    """
+
+    collections: tuple[str, ...]
+    """The collections to search.
+
+    Order matters if this dataset type is later referenced by a `FindFirst`
+    operation.  Collection wildcards are always resolved before being included
+    in a dataset search.
+    """
+
+    dimensions: DimensionGroup
+    """The dimensions of the dataset type.
+
+    This must match the dimensions of the dataset type as already defined in
+    the butler database, but this cannot generally be verified when a relation
+    tree is validated (since it requires a database query) and hence must be
+    checked later.
+    """
+
+    storage_class_name: str | None
+    """Name of the storage class to use when returning `DatasetRef` results.
+
+    May be `None` if the dataset is only used as a constraint or to return
+    columns that do not include a full dataset type.
+    """
```

Just FYI the strategy I've been using so far on StorageClass is that the client never sends the server anything about a StorageClass -- the server sends back a ref with the default storage class for the data type and the client overrides it if required.  This theoretically lets us support clients loading data using custom/experimental storage classes not known to the server.

I don't think it's harmful to have this get sent to the server and ignored, though at first glance I would have expected this bit of data to be attached to the SingleTypeDatasetQueryResults object and not in this part of the QueryTree... since the dataset types provided in the call to `datasets()` define the output from the query, not any intermediate joins.

---

I struggled with this a lot myself.  I added it because:
 - if `Query.join_dataset_search` can take a `DatasetType` and the user wants dataset results, it'll be confusing if they passed in a storage class override initially and then we forgot it just because we didn't have a place to put it;
 - I wanted `Query.join_dataset_search` to be able to take `DatasetType` as well as `str`, both so `Query.datasets` could delegate to it and so I didn't have to document why it was unusual in that respect.

I think the real problem is that `DatasetType` has a storage class, and that's a problem in lots of other places, but of course it's tremendously useful in probably even more other places and having a `DatasetTypeWithoutStorageClass` public class is its own problem.  And that's all thoroughly water under the bridge.

In any case, yes, I think the server can ignore this (or just blindly copy it into Parquet files, so the client doesn't have to edit them); the important thing is that the client doesn't forget it, and I think it's slightly better (even if that means sending it to the server) than out on its own in `Query` (and it would be `Query`, not `SingleTypeDatasetResults`, to serve its current purpose).

---

> if `Query.join_dataset_search` can take a `DatasetType` and the user wants dataset results, it'll be confusing if they passed in a storage class override initially and then we forgot it just because we didn't have a place to put it

Don't they already have to pass the dataset type again when they call `datasets()` to declare that they want refs back for this dataset type?  To me it makes sense that whatever you pass into `datasets` is the authoritative thing, since that's where you're declaring the output format.

>  not `SingleTypeDatasetResults`, to serve its current purpose).

It's possible I'm misunderstanding what is going on with this StorageClass (or maybe all the effects that a StorageClass can have?)  To me it looks like you could just pass in the StorageClass when you create the SingleTypeDatasetQueryResults, and call `ref.overrideStorageClass` if needed just before yielding from `__iter__`.  If this doesn't work I'd like to know because I'm using a similar pattern a few places in RemoteButler.

---

In almost all cases, that's right - it makes more sense to pass the storage class in later.  It's really quite an edge case I'm worried about here:
- user calls `Query.join_dataset_search` with a `DatasetType` instance that has a storage class that is not the same as the default one for that dataset type in the repository;
- user calls `Query.datasets` with a `str` dataset type name that has no storage class information.  Do we respect their previous storage class override?

Maybe it would be better to raise an exception or emit a warning if the `DatasetType` passed to `join_dataset_search` has a storage class that isn't the same as the repository one (since we need to look up the dataset type definition in the repository at that point anyway, to get the dimensions), saying that the storage class passed in there will be ignored.

---

Done on DM-43146: `DatasetSearch` no longer has `storage_class_name`.  Instead `join_dataset_search` raises if you give it a storage class override.